src/block_entity.zig

```diff
@@ -203,6 +226,263 @@ pub const BlockEntityTypes = struct {
+
+		pub fn renderAll(_: Mat4f, _: Vec3f, _: Vec3d) void {}
+	};
+
+	pub const Sign = struct {
+		const StorageServer = BlockEntityDataStorage(struct {
+			text: []const u8,
+		});
+		const StorageClient = BlockEntityDataStorage(struct {
+			text: []const u8,
+			renderedTexture: ?main.graphics.Texture = null,
+			blockPos: Vec3i,
+			block: main.blocks.Block,
+
+			fn deinit(self: @This()) void {
+				main.globalAllocator.free(self.text);
+				if(self.renderedTexture) |texture| {
+					textureDeinitLock.lock();
+					defer textureDeinitLock.unlock();
+					textureDeinitList.append(texture);
+				}
+			}
+		});
+		var textureDeinitList: main.List(graphics.Texture) = undefined;
+		var textureDeinitLock: std.Thread.Mutex = .{};
+		var pipeline: graphics.Pipeline = undefined;
+		var uniforms: struct {
+			ambientLight: c_int,
+			projectionMatrix: c_int,
+			viewMatrix: c_int,
+			playerPositionInteger: c_int,
+			playerPositionFraction: c_int,
+			quadIndex: c_int,
+			lightData: c_int,
+			chunkPos: c_int,
+			blockPos: c_int,
+		} = undefined;
+
+		// TODO: Load these from some per-block settings
+		const textureWidth = 128;
+		const textureHeight = 72;
+		const textureMargin = 4;
+
+		pub const id = "sign";
+		pub fn init() void {
+			StorageServer.init();
+			StorageClient.init();
+			textureDeinitList = .init(main.globalAllocator);
+
+			pipeline = graphics.Pipeline.init(
+				"assets/cubyz/shaders/block_entity/sign.vert",
+				"assets/cubyz/shaders/block_entity/sign.frag",
+				"",
+				&uniforms,
+				.{},
+				.{.depthTest = true, .depthCompare = .equal, .depthWrite = false},
+				.{.attachments = &.{.alphaBlending}},
+			);
+		}
+		pub fn deinit() void {
+			while(textureDeinitList.popOrNull()) |texture| {
+				texture.deinit();
+			}
+			textureDeinitList.deinit();
+			pipeline.deinit();
+			StorageServer.deinit();
+			StorageClient.deinit();
+		}
+		pub fn reset() void {
+			StorageServer.reset();
+			StorageClient.reset();
+		}
+
+		pub fn onUnloadClient(dataIndex: BlockEntityIndex) void {
+			StorageClient.mutex.lock();
+			defer StorageClient.mutex.unlock();
+			const entry = StorageClient.removeAtIndex(dataIndex) orelse unreachable;
+			entry.deinit();
+		}
+		pub fn onUnloadServer(dataIndex: BlockEntityIndex) void {
+			StorageServer.mutex.lock();
+			defer StorageServer.mutex.unlock();
+			const entry = StorageServer.removeAtIndex(dataIndex) orelse unreachable;
+			main.globalAllocator.free(entry.text);
+		}
+		pub fn onInteract(pos: Vec3i, chunk: *Chunk) EventStatus {
+			if(main.KeyBoard.key("shift").pressed) return .ignored;
+
+			StorageClient.mutex.lock();
+			defer StorageClient.mutex.unlock();
+			const data = StorageClient.get(pos, chunk);
+			main.gui.windowlist.sign_editor.openFromSignData(pos, if(data) |_data| _data.text else "");
+
+			return .handled;
+		}
+
+		pub const onLoadClient = updateClientData;
+		pub fn updateClientData(pos: Vec3i, chunk: *Chunk, reader: ?*BinaryReader) BinaryReader.AllErrors!void {
```

Comment:
That optional binary reader is enraging me. Devs are enraged.
What does it even mean. Apparently this is a delete update type. So please, make it an enum so I don't have to guess, not read the content of that function to determine what nullability means.

Reply:
Oh no, we can't have the devs enraged.